### DISCUSSION TITLE
How can I train on a dynamically generated dataset (based on the model’s own outputs) using GRPO Trainer?

### DISCUSSION THREAD

**Discussion by xtitx0327** [huggingface/trl#3542] [Q&A]
I’m new to RL and deep learning, so my question might seem simple. I would greatly appreciate any advice!

I’m using LLM to solve a multi-step sequence generation task. At each step:

- The model receives (a) a system prompt and (b) the current sequence generated so far.
- The model must choose one of three actions:

1.    Append something new to the sequence
2.    Remove/modify something already in the current sequence
3.    Terminate the generation process

I have implemented multiple reward functions to evaluate each action (append/remove/modify/terminate). Thus, the RL loop is:

1.    Model generates an action.
2.    Reward functions evaluate that action and assign a reward.
3.    Model updates itself based on these rewards.

Conceptually, this fits the GRPO training loop. The problem is, my training data is not a fixed (“static”) dataset—instead, it’s generated on‐the‐fly from the model’s own past outputs. According to #3213 , the current GRPO Trainer does not support IterableDataset.

**Question:** What’s the recommended way to handle a dynamically generated dataset with GRPO Trainer? Is there a workaround, or do I need to implement a custom training loop? Thank you for any pointers!

**Comment by The-Inscrutable-X** [comment]
I have the same question.

**Comment by akashcodes23** [comment]
> I’m new to RL and deep learning, so my question might seem simple. I would greatly appreciate any advice!
> 
> I’m using LLM to solve a multi-step sequence generation task. At each step:
> 
> * The model receives (a) a system prompt and (b) the current sequence generated so far.
> * The model must choose one of three actions:
> 
> 1. Append something new to the sequence
> 2. Remove/modify something already in the current sequence
> 3. Terminate the generation process
> 
> I have implemented multiple reward functions to evaluate each action (append/remove/modify/terminate). Thus, the RL loop is:
> 
> 1. Model generates an action.
> 2. Reward functions evaluate that action and assign a reward.
> 3. Model updates itself based on these rewards.
> 
> Conceptually, this fits the GRPO training loop. The problem is, my training data is not a fixed (“static”) dataset—instead, it’s generated on‐the‐fly from the model’s own past outputs. According to #3213 , the current GRPO Trainer does not support IterableDataset.
> 
> **Question:** What’s the recommended way to handle a dynamically generated dataset with GRPO Trainer? Is there a workaround, or do I need to implement a custom training loop? Thank you for any pointers!

Hi @xtitx0327,

Great question! This is a common pattern in RL for sequence generation tasks. I recently tackled a similar challenge in my HalluciSense project (LLM hallucination detection with iterative refinement).

## The Core Problem

GRPO Trainer currently expects a static `Dataset`, but your multi-step generation task requires dynamic, on-the-fly data generation where each step depends on previous model outputs.

## Recommended Workaround

Since `IterableDataset` isn't supported yet (as you noted in #3213), here's a practical workaround using a **custom data collator with step-wise generation**:

```python
from trl import GRPOTrainer, GRPOConfig
from datasets import Dataset
import torch

class DynamicStepDataset:
    def __init__(self, base_prompts, max_steps=10):
        self.base_prompts = base_prompts
        self.max_steps = max_steps
    
    def generate_step_data(self, model, batch_size=4):
        """Generate training data dynamically for one step"""
        # Your custom logic here:
        # 1. Sample current states (partial sequences)
        # 2. Model generates actions (append/remove/modify/terminate)
        # 3. Reward functions score each action
        # 4. Return formatted data for GRPO training
        
        dynamic_data = []
        for prompt in self.base_prompts[:batch_size]:
            # Simulate your sequence generation logic
            step_data = {
                "prompt": prompt,
                "completion": self.get_model_completion(model, prompt),
                "reward": self.calculate_reward(prompt)  # Your multi-reward logic
            }
            dynamic_data.append(step_data)
        
        return Dataset.from_list(dynamic_data)
    
    def get_model_completion(self, model, prompt):
        # Your generation logic
        pass
    
    def calculate_reward(self, prompt):
        # Your multi-reward function (append/remove/modify/terminate)
        pass

# Training loop with periodic data refresh
trainer = GRPOTrainer(
    model=model,
    reward_funcs=your_reward_function,
    args=GRPOConfig(output_dir="./grpo_dynamic", num_train_epochs=1),
    train_dataset=initial_dataset  # Start with static seed data
)

# Periodic re-generation approach
for step in range(num_training_rounds):
    # Train for a few steps on current data
    trainer.train(resume_from_checkpoint=True)
    
    # Generate new data based on updated model
    new_data = dynamic_dataset.generate_step_data(trainer.model)
    trainer.train_dataset = new_data  # Refresh dataset
```

Alternative: Custom Trainer Extension

If you need tighter integration, consider extending GRPOTrainer with a custom compute_loss that handles your step-wise generation internally, then feed pre-generated static batches to the parent class.
Key Considerations

Memory: Dynamic generation can be memory-intensive with large models—consider using accelerate for multi-GPU
Reward stability: Ensure your reward functions are stable across steps to prevent training drift
Checkpointing: Save intermediate states since dynamic generation makes resuming complex
Would this approach work for your use case? Happy to help refine it further!
Best,
Akash
